### PR TITLE
Fixes to hero / prison distribution

### PR DESCRIPTION
**Pull request by DjWarmonger**
- New class PrisonHeroDistributor
- Rolling and banning heroes is now protected with Mutex
- Do not iterate through entire hero list every time
- Misplaced or otherwise destroyed Prisons will return their heroes to the pool
- Same is true for Seer Huts which will restore quest artifacts to the pool and unban them on map, globally
- Quest artifacts on random map will correctly remain banned in game

### CODE REVIEW

**Comment by DjWarmonger** [comment]
Found an issue: Whenever Prison object is deleted (for instance due to failed treasure placement), hero stays banned. However, Prison pool size is restored so another Prison can spawn.

The same is true for quest artifacts.

**Comment by DjWarmonger** [comment]
Ready for review.
https://github.com/vcmi/vcmi/issues/3295 is still an issue, but this will need to wait for another day,

File: lib/rmg/CMapGenerator.h [hunk 1]
```diff
@@ -65,8 +65,8 @@ class DLL_LINKAGE CMapGenerator: public Load::Progress
 	const std::vector<ArtifactID> & getAllPossibleQuestArtifacts() const;
 	const std::vector<HeroTypeID> getAllPossibleHeroes() const;
 	void banQuestArt(const ArtifactID & id);
-	void banHero(const HeroTypeID& id);
-
+	void unbanQuestArt(const ArtifactID & id);
+
```
Comment (Alexander-Wilms) [inline lib/rmg/CMapGenerator.h:69]: Trailing whitespace
```suggestion

```

File: lib/rmg/modificators/PrisonHeroPlacer.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+private:
+    void getAllowedHeroes();
+	size_t reservedHeroes;
+
+protected:
+
+    std::vector<HeroTypeID> allowedHeroes;
+};
+
+VCMI_LIB_NAMESPACE_END
```
Comment (Alexander-Wilms) [inline lib/rmg/modificators/PrisonHeroPlacer.h:41]: Missing newline
```suggestion
VCMI_LIB_NAMESPACE_END

```

**Pull request merged**